### PR TITLE
[top] Add remaining TUSB1106 pins to CW310 top

### PR DESCRIPTION
**Pull request by a-will**
Set the SPEED pin on the TUSB1106 to full-speed (high) and the SUSPEND
pin to disable low-power mode (low). Previously, these pins relied on
the default pull-down behavior, which would have set the CW310's
on-board transceiver to low-speed mode.

Also constrain the port-to-register timing for the critical sampled
signals. The USB PHY's sampling logic expects to see at most one
transitive / fake SE0. Use set_max_delay to prevent excessive "P/N"
mismatch due to Vivado's whims on unconstrained paths.

Add register-to-port constraints to keep the USB outputs within the
full-speed transition time as well. Only the skew actually matters, but
the constraints use set_output_delay anyway.

Tested the hello_usbdev example with the CW310's on-board TUSB1106 PHY.

Signed-off-by: Alexander Williams <awill@google.com>

### CODE REVIEW

File: hw/top_earlgrey/data/clocks.xdc [hunk 1]
```diff
@@ -27,4 +27,14 @@ create_clock -add -name rv_jtag_tck -period 100.00 -waveform {0 5} [get_pin top_
...
 
+## USB input delay to accommodate T_FST and the PHY's sampling logic
+## The PHY expects to only see up to one transient / fake SE0. The phase
+## relationship with the PHY's sampling clock is arbitrary, but for simplicity,
+## constrain the maximum path delay to something smaller than
+## `T_sample - T_FST(max)` to help keep the P/N skew from slipping beyond one
+## sample period.
+set_max_delay -from [get_ports {IO_UPHY_DP_RX IO_UPHY_DN_RX IO_UPHY_D_RX}] -to ${clks_48_unbuf} 6
+set_output_delay -min -clock ${clks_48_unbuf} -3 [get_ports {IO_UPHY_DP_TX IO_UPHY_DN_TX IO_UPHY_OE_N}]
+set_output_delay -max -clock ${clks_48_unbuf} 4 [get_ports {IO_UPHY_DP_TX IO_UPHY_DN_TX IO_UPHY_OE_N}] -add
```
Comment (a-will) [inline hw/top_earlgrey/data/clocks.xdc:38]: Oops, looks like this file is shared with chip_englishbreakfast_cw305, and it doesn't have these ports. Hm... I guess I will kick this over to pins_cw310.xdc and pins_nexysvideo.xdc.

File: hw/top_earlgrey/data/pins_nexysvideo.xdc [hunk 2]
```diff
@@ -153,6 +153,17 @@ set_property -dict { PACKAGE_PIN T6    IOSTANDARD LVCMOS33 } [get_ports { IO_UPH
 set_property -dict { PACKAGE_PIN AB7   IOSTANDARD LVCMOS33 } [get_ports { IO_UPHY_OE_N }]; #IO_L20P_T3_34 Sch=jc_p[4]
 set_property -dict { PACKAGE_PIN AB6   IOSTANDARD LVCMOS33 } [get_ports { IO_UPHY_D_RX }]; #IO_L20N_T3_34 Sch=jc_n[4]
 
+## USB input delay to accommodate T_FST and the PHY's sampling logic
```
Comment (vogelpi) [inline hw/top_earlgrey/data/pins_nexysvideo.xdc:156]: Thanks for not forgetting about the NexysVideo!

File: hw/top_earlgrey/data/pins_cw310.xdc [hunk 3]
```diff
@@ -78,8 +78,19 @@ set_property -dict { PACKAGE_PIN AF14  IOSTANDARD LVCMOS18 } [get_ports { IO_UPH
 set_property -dict { PACKAGE_PIN P18   IOSTANDARD LVCMOS33 } [get_ports { IO_UPHY_SENSE }]; #USRUSB_VBUS_DETECT
 set_property -dict { PACKAGE_PIN AE15  IOSTANDARD LVCMOS18 } [get_ports { IO_UPHY_OE_N }]; #USRUSB_OE
 set_property -dict { PACKAGE_PIN V17   IOSTANDARD LVCMOS18 } [get_ports { IO_UPHY_D_RX }]; #USRUSB_RCV
-#set_property -dict { PACKAGE_PIN AE16  IOSTANDARD LVCMOS18 } [get_ports { IO_UPHY_SPD  }]; #USRUSB_SPD
-#set_property -dict { PACKAGE_PIN AF15  IOSTANDARD LVCMOS18 } [get_ports { IO_UPHY_SUS  }]; #USRUSB_SUS
+set_property -dict { PACKAGE_PIN AE16  IOSTANDARD LVCMOS18 } [get_ports { IO_UPHY_SPD  }]; #USRUSB_SPD
+set_property -dict { PACKAGE_PIN AF15  IOSTANDARD LVCMOS18 } [get_ports { IO_UPHY_SUS  }]; #USRUSB_SUS
+
+## USB input delay to accommodate T_FST and the PHY's sampling logic
```
Comment (vogelpi) [inline hw/top_earlgrey/data/pins_cw310.xdc:84]: What does T_FST stand for?

File: hw/top_earlgrey/data/pins_cw310.xdc [hunk 4]
```diff
@@ -78,8 +78,19 @@ set_property -dict { PACKAGE_PIN AF14  IOSTANDARD LVCMOS18 } [get_ports { IO_UPH
...
+## The PHY expects to only see up to one transient / fake SE0. The phase
+## relationship with the PHY's sampling clock is arbitrary, but for simplicity,
+## constrain the maximum path delay to something smaller than
+## `T_sample - T_FST(max)` to help keep the P/N skew from slipping beyond one
+## sample period.
+set clks_48_unbuf [get_clocks -of_objects [get_pin clkgen/pll/CLKOUT1]]
+set_max_delay -from [get_ports {IO_UPHY_DP_RX IO_UPHY_DN_RX IO_UPHY_D_RX}] -to ${clks_48_unbuf} 6
+set_output_delay -min -clock ${clks_48_unbuf} -3 [get_ports {IO_UPHY_DP_TX IO_UPHY_DN_TX IO_UPHY_OE_N}]
+set_output_delay -max -clock ${clks_48_unbuf} 4 [get_ports {IO_UPHY_DP_TX IO_UPHY_DN_TX IO_UPHY_OE_N}] -add
```
Comment (vogelpi) [inline hw/top_earlgrey/data/pins_cw310.xdc:93]: Shouldn't this be `-add_delay` instead of `-add`?

File: hw/top_earlgrey/data/pins_cw310.xdc [hunk 5]
```diff
@@ -78,8 +78,19 @@ set_property -dict { PACKAGE_PIN AF14  IOSTANDARD LVCMOS18 } [get_ports { IO_UPH
...
+
+## USB input delay to accommodate T_FST and the PHY's sampling logic
+## The PHY expects to only see up to one transient / fake SE0. The phase
+## relationship with the PHY's sampling clock is arbitrary, but for simplicity,
+## constrain the maximum path delay to something smaller than
+## `T_sample - T_FST(max)` to help keep the P/N skew from slipping beyond one
+## sample period.
+set clks_48_unbuf [get_clocks -of_objects [get_pin clkgen/pll/CLKOUT1]]
+set_max_delay -from [get_ports {IO_UPHY_DP_RX IO_UPHY_DN_RX IO_UPHY_D_RX}] -to ${clks_48_unbuf} 6
+set_output_delay -min -clock ${clks_48_unbuf} -3 [get_ports {IO_UPHY_DP_TX IO_UPHY_DN_TX IO_UPHY_OE_N}]
```
Comment (vogelpi) [inline hw/top_earlgrey/data/pins_cw310.xdc:92]: Question: is the negative value really required here? Or can we just take 0 and increase the max instead?

File: hw/top_earlgrey/data/pins_cw310.xdc [hunk 6]
```diff
@@ -78,8 +78,19 @@ set_property -dict { PACKAGE_PIN AF14  IOSTANDARD LVCMOS18 } [get_ports { IO_UPH
...
 set_property -dict { PACKAGE_PIN AE15  IOSTANDARD LVCMOS18 } [get_ports { IO_UPHY_OE_N }]; #USRUSB_OE
 set_property -dict { PACKAGE_PIN V17   IOSTANDARD LVCMOS18 } [get_ports { IO_UPHY_D_RX }]; #USRUSB_RCV
-#set_property -dict { PACKAGE_PIN AE16  IOSTANDARD LVCMOS18 } [get_ports { IO_UPHY_SPD  }]; #USRUSB_SPD
-#set_property -dict { PACKAGE_PIN AF15  IOSTANDARD LVCMOS18 } [get_ports { IO_UPHY_SUS  }]; #USRUSB_SUS
+set_property -dict { PACKAGE_PIN AE16  IOSTANDARD LVCMOS18 } [get_ports { IO_UPHY_SPD  }]; #USRUSB_SPD
+set_property -dict { PACKAGE_PIN AF15  IOSTANDARD LVCMOS18 } [get_ports { IO_UPHY_SUS  }]; #USRUSB_SUS
+
+## USB input delay to accommodate T_FST and the PHY's sampling logic
+## The PHY expects to only see up to one transient / fake SE0. The phase
+## relationship with the PHY's sampling clock is arbitrary, but for simplicity,
```
Comment (vogelpi) [inline hw/top_earlgrey/data/pins_cw310.xdc:86]: I am a bit confused here. Does the PHY use a sampling clock? I thought it's not clocked. At least in the data sheet I couldn't find any information on this.

File: hw/top_earlgrey/data/pins_cw310.xdc [hunk 7]
```diff
@@ -78,8 +78,19 @@ set_property -dict { PACKAGE_PIN AF14  IOSTANDARD LVCMOS18 } [get_ports { IO_UPH
...
+## constrain the maximum path delay to something smaller than
+## `T_sample - T_FST(max)` to help keep the P/N skew from slipping beyond one
+## sample period.
+set clks_48_unbuf [get_clocks -of_objects [get_pin clkgen/pll/CLKOUT1]]
```
Comment (vogelpi) [inline hw/top_earlgrey/data/pins_cw310.xdc:90]: This clock is already defined in `clocks.xdc`. Related to that, what is actually the right place for putting these I/O constraints? Any thoughts @msfschaffner , @tjaychen ?

**Review by vogelpi — COMMENTED** [review]
Hi @a-will ,
thanks for your PR! And sorry for the long delay, I completely forgot about this.

Your work looks very good. I just have a couple of questions regarding the constraints. And you would need to rebase the PR on top of current master. I think support for the suspend mode could be added in a follow-up PR. The signal is there in usbdev and going to the pinmux. We might need to adapt the pinmux config a bit.

File: hw/top_earlgrey/data/pins_cw310.xdc (hunk 3 above)
Comment (a-will) [reply hw/top_earlgrey/data/pins_cw310.xdc:84]: T_FST (or tFST, depending how you like to style it) is the maximum width of a false SE0. It's the "full-speed transition" time, beginning with the first data pin's transition and ending with the other one's. This is 14 ns in the spec.

File: hw/top_earlgrey/data/pins_cw310.xdc (hunk 6 above)
Comment (a-will) [reply hw/top_earlgrey/data/pins_cw310.xdc:86]: This is probably a difference in how we use terminology.

I think of the output of a PHY as being "raw" bits, so our PHY includes both the discrete transceiver (the asynchronous TUSB1106) and the sampling logic in the `usb_fs_nb_pe` module. These constraints are satisfying part of the Electrical section of the spec, which specifically identifies itself as the physical layer specification.

File: hw/top_earlgrey/data/pins_cw310.xdc (hunk 5 above)
Comment (a-will) [reply hw/top_earlgrey/data/pins_cw310.xdc:92]: As mentioned in the comment, the range is actually arbitrary--It just needs to reduce the skew between D+ and D- transitions to < 14 ns.

I'm not sure anymore why I went with modeling a hold time, though. Most likely, that was just as arbitrary, and I "centered" a 7 ns keep-out around the sampling clock. Going with min 0 and max 7 should also work (and keep the upstream port a little closer to the I/O buffers).

File: hw/top_earlgrey/data/pins_cw310.xdc (hunk 4 above)
Comment (a-will) [reply hw/top_earlgrey/data/pins_cw310.xdc:93]: Ah, I guess both are accepted, but you're right, `-add_delay` is what's in the documentation everywhere. Will fix. :)

File: hw/top_earlgrey/data/pins_cw310.xdc (hunk 7 above)
Comment (a-will) [reply hw/top_earlgrey/data/pins_cw310.xdc:90]: This doesn't actually define a clock here; it merely reuses the Tcl variable name. Do the Tcl-like variables have global scope across XDC files? Also, would it make sense for `clocks.xdc` to be renamed `timing.xdc` and include other timing constraints?

Comment (vogelpi) [reply hw/top_earlgrey/data/pins_cw310.xdc:90]: Ah yes, you are right. This doesn't define a clock, just the name. To be honest I don't know about the scope across files. I wouldn't rename the file to not disturb the flow too much at this point. We can still move the I/O contraints there if people think its the better place.

Comment (cdgori) [reply hw/top_earlgrey/data/pins_cw310.xdc:90]: > This doesn't actually define a clock here; it merely reuses the Tcl variable name. Do the Tcl-like variables have global scope across XDC files? Also, would it make sense for `clocks.xdc` to be renamed `timing.xdc` and include other timing constraints?

Speaking a bit for others (@msfschaffner / @tjaychen ) - I think the idea is to keep the clocks and the timing constraints separate files as there were use cases for applying/propagating clocks only in some flows.

File: hw/top_earlgrey/data/pins_cw310.xdc (hunk 3 above)
Comment (a-will) [reply hw/top_earlgrey/data/pins_cw310.xdc]: Added the "full-speed transition" bit in the comment.

File: hw/top_earlgrey/data/pins_cw310.xdc (hunk 5 above)
Comment (vogelpi) [reply hw/top_earlgrey/data/pins_cw310.xdc]: This sounds good, thanks.

File: hw/top_earlgrey/data/pins_cw310.xdc (hunk 6 above)
Comment (vogelpi) [reply hw/top_earlgrey/data/pins_cw310.xdc]: Ah thanks for clarifying. This makes sense.

File: hw/top_earlgrey/data/pins_cw310.xdc (hunk 7 above)
Comment (vogelpi) [reply hw/top_earlgrey/data/pins_cw310.xdc:98]: Yes, this sounds very reasonable. Thinking more about this, the board-specific pin constraints file also defines a use case. Depending on the board layout etc. these constraints will change. So, let's keep this here.

**Review by vogelpi — COMMENTED** [review]
Thanks for clarifying my questions and updating the comments @a-will . This PR looks good to me. However, the FPGA build currently fails to meet timing. As it was passing before, should be maybe go back to the negative minimum output delay?

Also, I would recommend rebasing this PR on top of current master as there was another PR touching some pins (#9405).

**Comment by a-will** [comment]
> Thanks for clarifying my questions and updating the comments @a-will . This PR looks good to me. However, the FPGA build currently fails to meet timing. As it was passing before, should be maybe go back to the negative minimum output delay?
> 
> Also, I would recommend rebasing this PR on top of current master as there was another PR touching some pins (#9405).

Haha, I guess the last rebase I did pulled it into negative slack. It built and passed timing and a basic test from commits about a few days back. I'll take a look!

**Comment by a-will** [comment]
> Thanks for clarifying my questions and updating the comments @a-will . This PR looks good to me. However, the FPGA build currently fails to meet timing. As it was passing before, should be maybe go back to the negative minimum output delay?
> 
> Also, I would recommend rebasing this PR on top of current master as there was another PR touching some pins (#9405).

So it was actually the input path that was showing a hold time failure -- the `set_max_delay` constraint was getting affected by source clock latency, causing a narrower window than I intended. Doing `-datapath_only` would get rid of the source clock latency, but it _also_ removes the destination clock latency, which we don't want.

For this round, I've switched the constraint to a `set_input_delay` against the destination clock and told the tool that the the source clock latency is in the input delay. This is still a little funky, since we don't have a great way to express that the phase relationship is arbitrary. The current values may also rely on the PLL placement, since the negative phase offset created by the feedback clock is making the hold time work out correctly.

(Alternatively, we could remove `-source_latency_included` and tweak the delay values to reposition the window instead. This would effectively cancel out the clock path delay up to the PLL output. But either way, we're just pushing an arbitrary sampling window around to match up with a mesochronous clock. In both cases, we can adjust the values if this accidentally makes setup or hold too tight--In my build, there was plenty of slack for hold time at > 2 ns.)

**Review by vogelpi — APPROVED** [review]
Thanks @a-will for rebasing the PR once more and getting also the input timing constraints right. This looks good to me and I am now merging the PR.

**Pull request merged**